### PR TITLE
APP-1068 Support web-oauth jwks file auth handler

### PR DESCRIPTION
**Pull request by AdamMagaluk**
This PR introduces rdk usage of the goutils web-oauth rpc handler. This handler accepts JWTs payloads as authentication payload in the goutils/rpc as part of the `Authorization: Bearer ...` gRPC metadata/ HTTP header.

The web-oauth handler is configured with a KeyProvider backed by a JSON Web Token Keys file. The hander looks up the public key based on the `kid` in the JWT Header and verifies the token. The JWT must have the RPC claims like `auth_md` as well as having a valid `aud` claim based on a provided list of support audiences.

The RDK config and Robot Config API supports passing an arbitrary JSON Web Token Keys file to the RDK. If the config has this auth handler config the rpc server will be configured with the web-oauth handler backed by the JWKs and audience list.

WIP companion PR for App: https://github.com/viamrobotics/app/pull/950 - This will generate the config for the auth handler in a similar way as the remotes.

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -530,10 +532,69 @@ type AuthConfig struct {
 	TLSAuthEntities []string            `json:"tls_auth_entities"`
 }
 
+// WebOauthConfig contains the structued AuthHandlerConfig for the CredentialsTypeOAuthWeb type.
```
Comment (edaniels) [inline config/config.go:535]: [nit] WebOAuthConfig

File: config/config.go [hunk 2]
```diff
@@ -530,10 +532,69 @@ type AuthConfig struct {
...
+
+	// on validation the JSONKeySet is validated and parsed into this field and can be used.
+	ValidatedKeySet jwks.KeySet `json:"-"`
+}
+
+// Validate returns true if the WebOAuth is valid. Ensures each key is valid and meets the required constraints.
+func (c *WebOauthConfig) Validate(path string) error {
+	jsonJWKs, err := json.Marshal(c.JSONKeySet)
+	if err != nil {
+		return utils.NewConfigValidationError(fmt.Sprintf("%s.config", path), errors.Wrap(err, "failed to marshal jwks"))
```
Comment (edaniels) [inline config/config.go:550]: I think it's `jwks`, not config for both errors here

File: config/config.go [hunk 3]
```diff
@@ -530,10 +532,69 @@ type AuthConfig struct {
...
+	}
+
+	keyset, err := jwks.ParseKeySet(string(jsonJWKs))
+	if err != nil {
+		return utils.NewConfigValidationError(fmt.Sprintf("%s.config", path), errors.Wrap(err, "failed to parse jwks"))
+	}
+
+	allowedKeyTypes := map[string]bool{
```
Comment (edaniels) [inline config/config.go:558]: pull out to vars outside of function

File: config/config.go [hunk 4]
```diff
@@ -530,10 +532,69 @@ type AuthConfig struct {
...
+		"RS256": true,
+		"RS384": true,
+		"RS512": true,
+	}
+
+	for i := 0; i < keyset.Len(); i++ {
+		// validate keys
+		key, ok := keyset.Get(i)
+		if !ok {
+			return utils.NewConfigValidationError(fmt.Sprintf("%s.config", path), errors.New("failed to parse jwks, missing index"))
```
Comment (edaniels) [inline config/config.go:572]: need correct field names

File: config/config.go [hunk 5]
```diff
@@ -530,10 +532,69 @@ type AuthConfig struct {
...
+		"RSA": true,
+	}
+
+	allowedAlgs := map[string]bool{
+		"RS256": true,
+		"RS384": true,
+		"RS512": true,
+	}
+
+	for i := 0; i < keyset.Len(); i++ {
```
Comment (edaniels) [inline config/config.go:568]: we should validate it's non-empty for security reasons

File: robot/web/web.go [hunk 6]
```diff
@@ -149,6 +151,30 @@ func (app *robotWebApp) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+// Two known auth handlers (LocationSecret, WebOauth).
+func isManagedAuthHandlers(handlers []config.AuthHandlerConfig) bool {
```
Comment (edaniels) [inline robot/web/web.go:155]: is -> has

File: robot/web/web.go [hunk 7]
```diff
@@ -149,6 +151,30 @@ func (app *robotWebApp) ServeHTTP(w http.ResponseWriter, r *http.Request) {
...
+		if h.Type == rutils.CredentialsTypeRobotLocationSecret {
+			hasLocationSecretHandler = true
+		}
+
+		if h.Type == oauth.CredentialsTypeOAuthWeb {
+			hasWebOAuthHandler = true
+		}
+	}
+
+	// During rollout of weboauth feature we need to support the app returning only the 1 location secret.
```
Comment (edaniels) [inline robot/web/web.go:168]: mark with a TODO(JIRA) here?

File: robot/web/web.go [hunk 8]
```diff
@@ -818,6 +844,17 @@ func (svc *webService) initAuthHandlers(listenerTCPAddr *net.TCPAddr, options we
 					handler.Type,
 					rpc.MakeSimpleMultiAuthHandler(authEntities, locationSecrets),
 				))
+			case oauth.CredentialsTypeOAuthWeb:
+				webOauthOptions := oauth.WebOAuthOptions{
+					AllowedAudiences: handler.WebOauthConfig.AllowedAudiences,
+					KeyProvider:      jwks.NewStaticJWKKeyProvider(handler.WebOauthConfig.ValidatedKeySet),
+					EntityVerifier: func(ctx context.Context, entity string) (interface{}, error) {
+						return map[string]interface{}{"email": entity}, nil
```
Comment (edaniels) [inline robot/web/web.go:852]: maybe a comment here as to why email is entity

**Review by edaniels — CHANGES_REQUESTED** [review]
Done with first round.

File: config/config.go (hunk 4 above)
Comment (AdamMagaluk) [reply config/config.go]: Done.

File: config/config.go (hunk 5 above)
Comment (AdamMagaluk) [reply config/config.go:575]: Done. Though an empty list does not provide any security concerns. It defaults to closed.

File: config/config.go (hunk 3 above)
Comment (AdamMagaluk) [reply config/config.go]: Done.

File: config/config.go (hunk 2 above)
Comment (AdamMagaluk) [reply config/config.go]: Done.

File: config/config.go (hunk 1 above)
Comment (AdamMagaluk) [reply config/config.go]: Done.

File: config/config_test.go [hunk 9]
```diff
@@ -732,3 +737,173 @@ ph2C/7IgjA==
...
+					},
+				},
+			},
+		}
+
+		err := config.Ensure(true)
+		test.That(t, err.Error(), test.ShouldContainSubstring, "failed to parse jwks")
+	})
+
+	t.Run("web-oauth handler valid config", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (AdamMagaluk) [reply config/config_test.go:825]: Added

File: robot/web/web.go (hunk 6 above)
Comment (AdamMagaluk) [reply robot/web/web.go:155]: done

File: robot/web/web.go (hunk 7 above)
Comment (AdamMagaluk) [reply robot/web/web.go:168]: done.

File: robot/web/web.go (hunk 8 above)
Comment (AdamMagaluk) [reply robot/web/web.go:852]: Added. I wanted email to be consistent with app. Once we start doing interesting things on the RDK for authorization etc.. we'll probably add a similar state structure like we have in the app.

I have a companion PR working now: https://github.com/viamrobotics/app/pull/950 the app will auto add to the AuthHandlers. Im open to debate if the RDK should be doing that or just accepting the auth handler as is. To be having the cloud edit the config makes it much more visible and easier to reason about then trying to understand the underlying logic in the RDK if managed or not.

Comment (edaniels) [reply robot/web/web.go:854]: I agree. I regret my past decisions :)

**Review by edaniels — APPROVED** [review]
LGTM

**Review by cheukt — APPROVED** [review]

File: robot/web/web.go (hunk 8 above)
Comment (AdamMagaluk) [reply robot/web/web.go:854]: Ack :)

**Pull request merged**